nifi-nar-bundles/nifi-stateless-processor-bundle/nifi-stateless-processor/src/main/java/org/apache/nifi/processors/stateless/ExecuteStateless.java

```diff
@@ -0,0 +1,1202 @@
+
+        final DataflowDefinition<VersionedFlowSnapshot> dataflowDefinition = createDataflowDefinition(context, flowSnapshot);
+
+        final StatelessDataflow dataflow = bootstrap.createDataflow(dataflowDefinition);
+        dataflow.initialize();
+        return dataflow;
+    }
+
+    private StatelessDataflow getDataflow(final ProcessContext context) throws IOException, StatelessConfigurationException {
+        final StatelessDataflow dataflow = dataflows.poll();
+        if (dataflow == null) {
+            return createDataflow(context);
+        }
+
+        return dataflow;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        // Fetch a FlowFile, if appropriate
+        FlowFile flowFile = null;
+        if (context.hasIncomingConnection()) {
+            flowFile = session.get();
+            if (flowFile == null) {
+                return;
+            }
+        }
+
+        // Get the dataflow to run
+        final StatelessDataflow dataflow;
+        try {
+            dataflow = getDataflow(context);
+        } catch (final Exception e) {
+            getLogger().error("Could not create dataflow from snapshot", e);
+            session.rollback();
+            return;
+        }
+
+        // Trigger the dataflow and make sure that we always add the StatelessDataflow object back to the queue so that it can be reused.
+        try {
+            runDataflow(dataflow, flowFile, context, session);
+        } finally {
+            dataflows.offer(dataflow);
+        }
+    }
+
+
+    private void runDataflow(final StatelessDataflow dataflow, final FlowFile flowFile, final ProcessContext context, final ProcessSession session) {
+        // Ensure that we get a legitimate timeout value
+        final long timeoutMillis;
+        try {
+            timeoutMillis = context.getProperty(DATAFLOW_TIMEOUT).evaluateAttributeExpressions(flowFile).asTimePeriod(TimeUnit.MILLISECONDS);
+        } catch (final Exception e) {
+            getLogger().error("Failed to determine Dataflow Timeout for {}. Routing to failure", flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        // Attempt to enqueue the dataflow. If unable, the appropriate log messages will be generated and actions taken by the enqueueFlowFile() method, so we can simply return.
+        if (flowFile != null) {
+            final boolean enqueued = enqueueFlowFile(flowFile, dataflow, context, session);
+            if (!enqueued) {
+                return;
+            }
+        }
+
+        // Reset any counters on the dataflow. This way, we can simply gather the counters after triggering the dataflow,
+        // and we know that's how much we need to adjust our counters by.
+        dataflow.resetCounters();
+
+        // Trigger the dataflow
+        final BulletinRepository bulletinRepository = dataflow.getBulletinRepository();
```

Comment:
I may be missing something, but I couldn't get any bulletins to appear on the ExecuteStateless processor when running a dataflow that produced an ERROR bulletin -- `maxBulletinId` was always -1 here during my testing.

Reply:
Thanks. You're right. Those were not properly surfacing. I had to make a few adjustments.